### PR TITLE
Facility details beneath each District

### PR DESCRIPTION
**Pull request by claudiovallejo**
**Story card:** https://app.clubhouse.io/simpledotorg/story/435/facility-details-beneath-each-district

## Because

Enter the reason for raising this PR...

## This addresses
* "HTN controlled" facilities table
* "No BP measure" facilities table
* "HTN not under control" facilities table
* "Cumulative registrations" table

image

### CODE REVIEW

File: app/assets/javascripts/application.js [hunk 1]
```diff
@@ -27,6 +27,22 @@
 
 $(function () {
   $('[data-toggle="tooltip"]').tooltip()
+  
+  if($('#htn-controlled-table').length) {
```
Comment (kitallis) [inline app/assets/javascripts/application.js:31]: Let's pull these table names out into constants and just use them in both the `if` check and the actual element fetch.

Comment (kitallis) [reply app/assets/javascripts/application.js:31]: I think we can pull these selectors out into the specific page they refer to like `reports.js` or `user_analytics.js` etc.

**Comment by kitallis** [comment]
Other than some minor JS restructuring, looks good to me 👍

File: app/assets/javascripts/application.js (hunk 1 above)
Comment (claudiovallejo) [reply app/assets/javascripts/application.js]: Thanks @kitallis, I went ahead and moved all reports-specific selectors to `reports.js`. I'll create a separate story and PR to move analytics table selector to `user_analytics.js`.

File: app/assets/javascripts/reports.js [hunk 2]
```diff
@@ -1,14 +1,42 @@
...
+
+function initializeTables() {
+  const tableSortConfig = { descending: true };
+
+  const cumulativeRegistrationsTable = document.getElementById("cumulative-registrations-table");
+  const htnNotUnderControlTable = document.getElementById("htn-not-under-control-table");
+  const noBPMeasureTable = document.getElementById("no-bp-measure-table");
+  const htnControlledTable = document.getElementById("htn-controlled-table");
+
+  if (htnControlledTable) {
```
Comment (kitallis) [inline app/assets/javascripts/reports.js:24]: We seem to be doing this kind of thing a lot, in other places as well, and [not just tablesort](https://github.com/simpledotorg/simple-server/blob/master/app/assets/javascripts/user_analytics.js#L5) I could suggest an initialize function for things like this,

```javascript
function initElementById(element, func) { 
    if (document.getElementById(element)) { 
        func(); 
    } 
}

const htnControlTables = "htn-control-table"
const noBPMeasureTable = "no-bp-measure-table"

initElementById(htnControlTables,  () => new Tablesort(htnControlTables, tableSortConfig))
initElementById(noBPMeasureTable,  () => new Tablesort(noBPMeasureTable, tableSortConfig))
```

Probably not worth it for now though, maybe we can do this later.

**Comment by rsanheim** [comment]
Now that we have the bulk caching in place, listing all the facility details should be doable.

I also added caching for the NoBPMeasure queries...this time doing it at the per-period level instead of for an entire range of values.  The trade off here is that other callers who need smaller or larger result sets will hit the cache, with the downside being we are going to be way chattier w/ Redis. We could use a multi fetch type thing for Redis to pretty easily clean up that piece up though.

File: app/services/no_bp_measure_service.rb [hunk 3]
```diff
@@ -19,9 +22,15 @@ def call
 
   def visited_without_bp_taken_count(period)
     return 0 if facilities.empty?
+    Rails.cache.fetch(cache_key(period), version: cache_version, expires_in: CACHE_TTL, force: force_cache?) do
+      execute_sql(period)
+    end
+  end
+
+  def execute_sql(period)
```
Comment (rsanheim) [inline app/services/no_bp_measure_service.rb:30]: alternative approach to https://github.com/simpledotorg/simple-server/pull/1260

**Pull request merged**